File: platforms/glassdoor.py
"""Glassdoor India — Google login, job search, apply automation."""
import time
import random

try:
    from selenium.webdriver.common.by import By
    from selenium.webdriver.support.ui import WebDriverWait
    from selenium.webdriver.support import expected_conditions as EC
    from selenium.common.exceptions import TimeoutException, NoSuchElementException
    _SELENIUM_OK = True
except ImportError:
    _SELENIUM_OK = False  # Vercel — browser methods won't be called

from core.browser import get_driver, human_delay, safe_click
from core.config import Config
from core.logger import get_logger
from platforms.base import BasePlatform, JobListing

log = get_logger("glassdoor")
BASE = "https://www.glassdoor.co.in"


class GlassdoorPlatform(BasePlatform):
    name = "glassdoor"

    def __init__(self):
        self.driver = None

    def _init(self):
        if not self.driver:
            self.driver = get_driver("glassdoor")

    # ── Authentication ──────────────────────────────────────────
    def login(self) -> bool:
        self._init()
        self.driver.get(f"{BASE}/Job/index.htm")
        human_delay(3, 5)

        if self._logged_in():
            log.info("Glassdoor: already authenticated")
            return True

        log.info("Glassdoor: attempting Google login")
        self.driver.get(f"{BASE}/profile/login_input.htm")
        human_delay(2, 4)

        # Close any modal first
        self._close_modal()

        for selector in [
            "//a[contains(@href,'google')]",
            "//button[contains(.,'Google')]",
            "//*[contains(@data-brandviews,'google')]",
            "//div[contains(@class,'google')]//button",
        ]:
            try:
                btn = WebDriverWait(self.driver, 6).until(
                    EC.element_to_be_clickable((By.XPATH, selector))
                )
                safe_click(self.driver, btn)
                human_delay(2, 4)
                break
            except (TimeoutException, NoSuchElementException):
                continue

        self._handle_google_popup()

        for _ in range(40):
            human_delay(2, 2)
            if self._logged_in():
                log.info("Glassdoor: Google login successful")
                return True

        log.warning("Glassdoor: login not complete — run setup_login.py")
        return False

    def _logged_in(self) -> bool:
        try:
            self.driver.find_element(By.CSS_SELECTOR,
                ".SVGInline.avatar, [data-test='user-menu'], .userMenu, .dropdown-avatar")
            return True
        except NoSuchElementException:
            url = self.driver.current_url
            return "glassdoor" in url and "login" not in url and "signin" not in url

    def _handle_google_popup(self):
        human_delay(2, 3)
        windows = self.driver.window_handles
        if len(windows) > 1:
            self.driver.switch_to.window(windows[-1])
            human_delay(1, 2)
            try:
                account_divs = self.driver.find_elements(By.CSS_SELECTOR, "[data-identifier]")
                for div in account_divs:
                    if Config.CANDIDATE_EMAIL.lower() in (div.get_attribute("data-identifier") or "").lower():
                        safe_click(self.driver, div)
                        human_delay(2, 3)
                        break
            except Exception:
                pass
            self.driver.switch_to.window(windows[0])

    def _close_modal(self):
        for sel in ["[alt='Close']", "button.modal_closeIcon", "button[class*='close']", "[data-test='close']"]:
            try:
                self.driver.find_element(By.CSS_SELECTOR, sel).click()
                human_delay(0.5, 1)
                return
            except NoSuchElementException:
                continue

    # ── Job Search ───────────────────────────────────────────────
    def search_jobs(self, role: str, location: str) -> list[JobListing]:
        self._init()
        jobs: list[JobListing] = []

        url = (f"{BASE}/Job/jobs.htm"
               f"?sc.keyword={role.replace(' ', '+')}"
               f"&locT=N&locId=115&fromAge=7"
               f"&jobType=fulltime")
        self.driver.get(url)
        human_delay(3, 5)
        self._close_modal()
        self._scroll()

        cards = self.driver.find_elements(By.CSS_SELECTOR,
            "li.react-job-listing, [data-test='jobListing'], article[class*='job']")
        log.info(f"Glassdoor: {len(cards)} results for '{role}' @ {location}")

        for card in cards[:30]:
            try:
                job = self._parse_card(card)
                if job:
                    jobs.append(job)
            except Exception:
                continue
        return jobs

    def _parse_card(self, card) -> JobListing | None:
        try:
            title_el = card.find_element(By.CSS_SELECTOR,
                "a[data-test='job-title'], .jobLink, a[class*='jobTitle']")
            try:
                company_el = card.find_element(By.CSS_SELECTOR,
                    "[data-test='employer-short-name'], .jobHeader .employerName")
                company = company_el.text.strip()
            except NoSuchElementException:
                company = ""
            try:
                loc_el = card.find_element(By.CSS_SELECTOR,
                    "[data-test='emp-location'], .jobHeader .subtle")
                location = loc_el.text.strip()
            except NoSuchElementException:
                location = ""

            href = title_el.get_attribute("href") or ""
            if not href:
                return None
            if not href.startswith("http"):
                href = BASE + href

            return JobListing(
                platform="glassdoor",
                title=title_el.text.strip(),
                company=company,
                location=location,
                job_url=href.split("?")[0],
            )
        except NoSuchElementException:
            return None

    def _get_job_description(self, job: JobListing) -> str:
        try:
            self.driver.get(job.job_url)
            human_delay(2, 3)
            self._close_modal()
            desc = WebDriverWait(self.driver, 8).until(
                EC.presence_of_element_located((By.CSS_SELECTOR,
                    "[data-test='jobDescriptionContent'], .desc, .jobDescriptionContent"))
            )
            return desc.text[:3000]
        except Exception:
            return ""

    # ── Apply ────────────────────────────────────────────────────
    def apply_to_job(self, job: JobListing) -> bool:
        self._init()
        self.driver.get(job.job_url)
        human_delay(2, 4)
        self._close_modal()

        for selector in [
            "[data-test='applyButton']",
            "button[class*='apply']",
            "a[class*='apply']",
        ]:
            try:
                btn = WebDriverWait(self.driver, 6).until(
                    EC.element_to_be_clickable((By.CSS_SELECTOR, selector))
                )
                safe_click(self.driver, btn)
                human_delay(1, 2)
                log.info("Glassdoor: apply clicked", extra={"url": job.job_url})
                return True
            except TimeoutException:
                continue

        log.warning("Glassdoor: apply button not found", extra={"url": job.job_url})
        return False

    def _scroll(self):
        for _ in range(5):
            self.driver.execute_script("window.scrollBy(0, 700)")
            time.sleep(random.uniform(0.4, 0.9))

    def close(self):
        if self.driver:
            try:
                self.driver.quit()
            except Exception:
                pass
            self.driver = None

File: playwright/platforms/glassdoor.js
const { chromium } = require('playwright');
const path = require('path');
const fs = require('fs');
const config = require('../config');

const DELAY = (min, max) => new Promise(r => setTimeout(r, min + Math.random() * (max - min)));
const BASE = 'https://www.glassdoor.co.in';

function parseExpMin(expStr) {
  if (!expStr) return 0;
  const m = expStr.match(/(\d+)/);
  return m ? parseInt(m[1]) : 0;
}

class Glassdoor {
  constructor() {
    this.authFile = path.join(config.authDir, 'glassdoor_state.json');
    this.browser = null;
    this.context = null;
    this.page = null;
  }

  async init() {
    const storageState = fs.existsSync(this.authFile) ? this.authFile : undefined;
    this.browser = await chromium.launch({
      headless: config.headless,
      slowMo: 50,
      args: ['--disable-blink-features=AutomationControlled'],
    });
    this.context = await this.browser.newContext({
      storageState,
      viewport: { width: 1280, height: 800 },
      userAgent: 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
    });
    this.page = await this.context.newPage();
  }

  async isLoggedIn() {
    try {
      await this.page.goto(`${BASE}/Job/index.htm`, { timeout: 20000, waitUntil: 'domcontentloaded' });
      await DELAY(2000, 3000);
      await this.page.waitForSelector(
        '[data-test="user-menu"], .dropdown-avatar, .userMenu, .SVGInline.avatar',
        { timeout: 6000 }
      );
      return true;
    } catch {
      try {
        const url = this.page.url();
        return url.includes('glassdoor') && !url.includes('login') && !url.includes('signin');
      } catch {
        return false;
      }
    }
  }

  async closeModal() {
    for (const sel of [
      '[alt="Close"]',
      'button.modal_closeIcon',
      'button[class*="close"]',
      '[data-test="close"]',
      'button[aria-label*="close" i]',
    ]) {
      try {
        const btn = await this.page.$(sel);
        if (btn && await btn.isVisible()) { await btn.click(); await DELAY(400, 700); return; }
      } catch {}
    }
  }

  async searchJobs(role, location) {
    const keyword = encodeURIComponent(role);
    const url = `${BASE}/Job/jobs.htm?sc.keyword=${keyword}&locT=N&locId=115&fromAge=7&jobType=fulltime`;

    try {
      await this.page.goto(url, { timeout: 30000, waitUntil: 'domcontentloaded' });
      await DELAY(2500, 4000);
    } catch {
      return [];
    }

    await this.closeModal();

    for (let i = 0; i < 4; i++) {
      await this.page.evaluate('window.scrollBy(0, 700)');
      await DELAY(500, 900);
    }

    const cards = await this.page.$$(
      'li.react-job-listing, [data-test="jobListing"], article[class*="job"], li[data-id]'
    );
    const jobs = [];

    for (const card of cards.slice(0, 30)) {
      try {
        const titleEl = await card.$('a[data-test="job-title"], .jobLink, a[class*="jobTitle"]');
        if (!titleEl) continue;

        const title = (await titleEl.textContent() || '').trim();
        let href = (await titleEl.getAttribute('href') || '').split('?')[0];
        if (!href) continue;
        if (!href.startsWith('http')) href = BASE + href;

        const company = await card.$eval(
          '[data-test="employer-short-name"], .jobHeader .employerName, [class*="employer-name"]',
          el => el.textContent.trim()
        ).catch(() => '');

        const location = await card.$eval(
          '[data-test="emp-location"], .jobHeader .subtle, [class*="location"]',
          el => el.textContent.trim()
        ).catch(() => '');

        const exp = await card.$eval(
          '[data-test="job-meta"], [class*="job-age"], [class*="meta"]',
          el => el.textContent.trim()
        ).catch(() => '');

        jobs.push({ title, company, location, experience: exp, url: href });
      } catch {}
    }

    return jobs;
  }

  async applyToJob(job) {
    try {
      await this.page.goto(job.url, { timeout: 30000, waitUntil: 'domcontentloaded' });
      await DELAY(2000, 3500);
      await this.closeModal();

      const content = await this.page.content();
      if (content.includes('Applied') || content.includes('application submitted')) return false;

      let applyBtn = null;
      for (const sel of [
        '[data-test="applyButton"]',
        'button[class*="apply"]',
        'a[class*="apply"]',
        'button:has-text("Apply")',
      ]) {
        applyBtn = await this.page.$(sel);
        if (applyBtn && await applyBtn.isVisible()) break;
        applyBtn = null;
      }

      if (!applyBtn) return false;

      const pages = this.context.pages();
      await applyBtn.click();
      await DELAY(2000, 3500);

      // If new tab opened (external apply), close it
      const nowPages = this.context.pages();
      if (nowPages.length > pages.length) {
        for (const p of nowPages) {
          if (!pages.includes(p)) await p.close();
        }
        return false;
      }

      return true;
    } catch {
      return false;
    }
  }

  async run(target) {
    const result = { applied: 0, failed: 0, skipped: 0, experience_counts: {}, jobs: [] };

    await this.init();

    if (!await this.isLoggedIn()) {
      console.log('[Glassdoor] Not logged in. Run: node auth_setup.js first');
      await this.browser.close();
      return result;
    }
    console.log('[Glassdoor] Logged in ✓');

    outer: for (const role of config.roles) {
      for (const location of config.locations) {
        if (result.applied >= target) break outer;

        process.stdout.write(`[Glassdoor] ${role} @ ${location} → `);
        const jobs = await this.searchJobs(role, location);
        console.log(`${jobs.length} found`);

        for (const job of jobs) {
          if (result.applied >= target) break outer;

          const exp = job.experience || 'N/A';
          result.experience_counts[exp] = (result.experience_counts[exp] || 0) + 1;

          // Skip jobs requiring more experience than user has
          if (parseExpMin(exp) > config.maxExpYears) {
            console.log(`  ↳ [${exp}] ${job.title} @ ${job.company} ... skip (over-experienced)`);
            result.skipped++;
            continue;
          }

          process.stdout.write(`  ↳ [${exp}] ${job.title} @ ${job.company} ... `);
          const ok = await this.applyToJob(job);

          if (ok) {
            result.applied++;
            result.jobs.push({ ...job, status: 'applied', platform: 'glassdoor' });
            console.log('✓ Applied');
          } else {
            result.skipped++;
            console.log('skip');
          }

          await DELAY(1500, 2800);
        }
      }
    }

    await this.context.storageState({ path: this.authFile });
    await this.browser.close();
    return result;
  }
}

module.exports = Glassdoor;
